### ISSUE TITLE
[BUG][CML][UCM2]No sound output after unplug headset then plug again

### ISSUE THREAD

**Issue by Liviali155** [thesofproject/linux#1991]
Describe the bug
No sound output after unplugin headset then plugin again
Ubuntu 20.04
```
ubuntu@cml-hda-mantis-1:~$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofhdadsp [sof-hda-dsp], device 0: HDA Analog (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 1: HDA Digital (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 3: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 4: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofhdadsp [sof-hda-dsp], device 5: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
To Reproduce
1.sudo reboot
2.wait runtime pm status suspend
3.plugin headset
4.open sound settings
5.set input device as “Headset Mono Microphone-sof-hda-dsp”,output Device as "Headphone-sof-hda-dsp"
6.aplay -Dhw:0,0 -r48000 -c2 -fs16_le  wav/48California_Dreaming.wav -vv -i(has sound output) or paplay -vv wav/48California_Dreaming.wav(has sound output)
7.unplug headset
8.plugin headset again
9.has no sound output in headset

Reproduction Rate
10/10

Expected behavior
There is sound output after unplug headset then plugin again

Actual result
There is no sound output after unplug headset then plugin again
At the same time open sound settings,Input Device is Headphones Stereo Microphone-sof-hda-dsp
After set input device as “Headset Mono Microphone-sof-hda-dsp” , sound output normal

![565724532](https://user-images.githubusercontent.com/56707932/78663779-50771400-7905-11ea-9a7c-bbc7c9b6dd47.jpg)

Platform: CML laptop with onboard codec ALC3204 in HDA mode
Firmware: b4e8b1e4 https://github.com/thesofproject/sof/commits/master
Kernel: 052e45d7 https://github.com/thesofproject/linux/commits/topic/sof-dev
Topology: file: tools/topology/sof-hda-generic-2ch.tplg same as firmware
UCM2:UCM2:https://github.com/thesofproject/alsa-ucm-conf/tree/master/ucm2/sof-hda-dsp

[dmesg0407.log](https://github.com/thesofproject/linux/files/4443848/dmesg0407.log)
[sof-logger.log](https://github.com/thesofproject/linux/files/4443849/sof-logger.log)

**Comment by juimonen** [comment]
@Liviali155 So if after step 7 you set manually again the "mono mic", does it work? We can't too much rely on the mic detection if it is suppose to happen with the additional Dell pop-up...

So if the stereo mic is selected for some reason (as in this case), it will block the sound output (because it assumes there is jack with 2 mics connected, no headphones, the name is really misleading)

**Comment by Liviali155** [comment]
> @Liviali155 So if after step 7 you set manually again the "mono mic", does it work? We can't too much rely on the mic detection if it is suppose to happen with the additional Dell pop-up...
> 
> So if the stereo mic is selected for some reason (as in this case), it will block the sound output (because it assumes there is jack with 2 mics connected, no headphones, the name is really misleading)

After step7 unplug headset,there is no "“Headset Mono Microphone-sof-hda-dsp” item in input Device,It only has one item"Digital Microphone-sof-hda-dsp"

**Comment by Liviali155** [comment]
This issue also can be reproduced on sof v1.5-rc1
Kernel:4b4681eb Sof-dev
fw:5f7d4e2 https://github.com/thesofproject/sof/releases/tag/v1.5-rc1
UCM2(https://github.com/thesofproject/alsa-ucm-conf/tree/master/ucm2/sof-hda-dsp)

**Comment by juimonen** [comment]
@Liviali155 i have just default ubuntu 20, no updates. For me it seems that with first insert it works, as with the description. But when reinserting, because you have selected "headset mono mic" it will select "headphones stereo mic" -> this will cut the sound from headphones. So for some reason "headphones stereo mic" is preferred over "headset mono mic". This would really need the ui pop-up to do the selection.

We could try to figure out also that "headset mono" would be preferred over "headphone stereo" in input, but still we would need the pop-up.

**Comment by juimonen** [comment]
@Liviali155 it helps to reverse the order of "headphone mic" and "headset" in HiFi.conf in sof-hda-dsp. So I think pulseaudio chooses the first definition as primary without pop-up. At least for me the settings ui show now always headset mono mic when re-inserting the headset.

However I still get sometimes headset silenced. Alsamixer shows that the "Input Source" would be correct but I can't hear anything. toggling the input switch gets the audio audible. Also removing and inserting the headset sometimes helps. And this is with the paplay playing all the time.

**Comment by juimonen** [comment]
and it could be that the above is pulseaudio bug that @libinyang was reporting, with enable and disable sequences.

So we should re-test this bug with
1) reverse the order of "headphone mic" and "headset" definitions in HiFi.conf in sof-hda-dsp
2) adding the https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/276 to pulseaudio (already in fedora)

**Comment by juimonen** [comment]
I tried adding the pulseaudio patches above -> no help for the random headset mute

I did "cat /proc/asound/card0/codec#0 | grep 0x23 -A2" (0x23 should be the input source switch)
If I change the input source manually from alsamixer between "headset mic" and "headphone mic", I can see that above "cat" produces differences: "headset mic" second value pair 0, "headphone mic" third value pair zero. And also if doing playback, using the alsamixer Input Source switch will either cut the playback or make it audible. This works as expected.

When I get random muting in headset after remove/insert, alsamixer Input Source setting is correct (headset mic) and also "cat" command above produces correct output (second value pair 0). But everything is muted. Now if I toggle the Input Source switch back and forth I get audio back.

So something is fishy in the codec setup. I don't think pulseaudio/alsa will actually touch the Input Source as it is in the correct position already (when inserting the headset). But somehow the codec settings inside it are wrong...? In this use case there is no suspend/resume just insert event, when speakers are muted and headphones unmuted.

@kailangyang could you also check this and possibly give some hints? I don't see issues in pulseaudio/alsa/sof, somehow codec is in strange state.

**Comment by plbossart** [comment]
> 1. reverse the order of "headphone mic" and "headset" definitions in HiFi.conf in sof-hda-dsp

why would use the order matter?

**Comment by juimonen** [comment]
@plbossart seems that pulseaudio picks up the first one "mic" defined in this use case. There's 2 definitions in this case. So as a default it is much better to just put the "headset mic" as first item, because that is 99% of the cases...

Not sure if it can be made explicit with some priorities etc. And anyway this selection should be made with the pop-up.

**Comment by plbossart** [comment]
> @plbossart seems that pulseaudio picks up the first one "mic" defined in this use case. There's 2 definitions in this case. So as a default it is much better to just put the "headset mic" as first item, because that is 99% of the cases...
> 
> Not sure if it can be made explicit with some priorities etc. And anyway this selection should be made with the pop-up.

ah sorry, I read 'headphone' and 'headset' and couldn't figure out why playback and capture would be confused.

What is a 'headphone mic'?  that looks confusing to me, and yes priorities should be applied. If not PulseAudio is broken if it selects the first one.

**Comment by juimonen** [comment]
@plbossart "headphone mic" is basically a stereo analog mic. And the name, I guess, implies it is a "pure" mic stuck to headphone input. Dell folks want to have this I guess for some legacy reasons. 
And if you want to use this kind of device in headphone input, you need to configure the pins so that for sure there's no audio coming out. So if this kind of use case is loaded -> no audio out.

The name is really confusing, the device itself is quite rare, I don't personally remember ever seen this kind of device nor using it.  But that's just me.

Yeah, I can experiment a bit with priorities, but the quick medicine is just to reverse the ordering.

**Comment by plbossart** [comment]
thanks @juimonen so that's just a 3.5mm mic, nothing to do with headphones other than using a 3.5mm jack.

**Comment by juimonen** [comment]
@Liviali155 can you try https://github.com/alsa-project/alsa-ucm-conf/pull/21

**Comment by Liviali155** [comment]
> @Liviali155 can you try [alsa-project/alsa-ucm-conf#21](https://github.com/alsa-project/alsa-ucm-conf/pull/21)

Tried with alsa-project/alsa-ucm-conf#21,input device is “Headset Mono Microphone-sof-hda-dsp”  after unplug/plug headset,but sometimes still has no sound output in headset,need to change input device as "Headphones Stereo Microphone-sof-hda-dsp",then “Headset Mono Microphone-sof-hda-dsp” again,there is sound output in headset
log and amixer contents as attachment, seems there is no difference betweem amixercontents_pass.txt and amixercontents_fail.txt

[amixercontents_fail.txt](https://github.com/thesofproject/linux/files/4501375/amixercontents_fail.txt)
[amixercontents_pass.txt](https://github.com/thesofproject/linux/files/4501377/amixercontents_pass.txt)
[dmesg0420.log](https://github.com/thesofproject/linux/files/4501379/dmesg0420.log)
[sof-logger-0420.log](https://github.com/thesofproject/linux/files/4501380/sof-logger-0420.log)

**Comment by juimonen** [comment]
@Liviali155 yes I also noticed the same. I think that's the main issue now. I think this bug should be closed as we can get rid of it by ucm mic ordering or by the actual pop-up. And we should maybe create a new bug from above. 

You can get rid of it by toggling the "input Source" in alsamixer. So this might also actually go away with pop-up. And there's again something strange happening with this coded as alsamixer reports something else than what is the actual situation under the hood...

**Comment by Liviali155** [comment]
Because apllied alsa-project/alsa-ucm-conf#21,input device is “Headset Mono Microphone-sof-hda-dsp” after unplug/plug headset,about the issue  sometimes still has no sound output in headset has creat another issue to follow up https://github.com/thesofproject/linux/issues/2025,closed this issue

**Comment by ClarexZhou** [comment]
Reopen for might not merge Pr#21, wait pulseaudio update https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/294

**Comment by Liviali155** [comment]
On matins test result as below:
default pulseaudio of Ubuntu20.04+https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/294
HAD UCM2: https://github.com/thesofproject/alsa-ucm-conf/commits/master (commit: 81465c56) :
Input Device will audo switched to Headset when plug headset again with sound setting closed,and sound in headset normal when plug/unplug headset,but if open soundsettig,Input Device will not auto switched to headset when plug headset again,still is “Digital Microphone-sof-had-dsp”,but the sound in headset still normal

default pulseaudio of Ubuntu20.04+https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/294+ https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/295 
HAD UCM2: https://github.com/thesofproject/alsa-ucm-conf/commits/master (commit: 81465c56):

error not seen

**Comment by ClarexZhou** [comment]
Error not seen with https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/294 and  https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/295#note_512131, wait Pulseaudio update.

**Comment by kv2019i** [comment]
@libinyang I assigned this to you, to follow-up the PA merge request from SOF team.

**Comment by ClarexZhou** [comment]
Close for https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/merge_requests/294 is merged.
